docs/design/features/readytorun-elf-support-roadmap.md

```diff
@@ -0,0 +1,170 @@
+  using the COM descriptor directory entry in the PE header.
+
+* **Crossgen2 compiler**: The SectionBuilder and R2RPEBuilder needs to be
+  decoupled from the PE format by some form of virtualization. We need to
+  port the various ELF header structures to managed code for use in Crossgen2
+  and R2RDump (do we need some approval or special arrangements for porting
+  the Linux header files like elf.h and elftypes.h to managed code in the runtime
+  repo?). After we have the constants, structures and their serialization
+  mechanisms in place and after we figure out the proper abstraction the work
+  should be pretty straightforward in terms of just filling in the various header
+  and section tables and emitting the section data in the proper place.
+
+* **Diagnostics / SOS**: switching over to native ELF will have fundamental effect on
+  diagnostics. While the overall long-term effect is assumed to be positive,
+  it will take time to bridge the user experience gap. It is likely that for some
+  temporary period of time we'll need to keep support for both PE files and ELF
+  files on Linux until all diagnostic functionality for ELF has been put on par
+  with pre-existing diagnostic support for PE-based executables.
+
+* **Visual Studio, VS Code**: switching over the executable format is certain to
+  have non-trivial effect on both VS and VS Code. We need to plan ahead with the
+  relevant teams to make sure we cover all the relevant project template, build
+  system, debugger, profiler and other aspects.
+
+* **R2RDump**: The PE header reader that is today using the managed assembly
+  System.Reflection.PortableExecutable needs to be generalized to recognize and
+  parse the ELF headers and locate the RTR_HEADER export symbol - once we're able
+  to identify the R2R header in the file, we immediately have access to all the
+  sections and data. Naturally we'll also need to adjust the current logic for
+  translating RVA's to file offsets.
+
+* **ILSpy**: will hopefully continually ingest R2RDump changes so it should
+  automatically pick up support for ELF-embedded MSIL/ECMA-335 metadata.
+
+# Proposed sequencing
+
+The natural sequencing and initial T-shirt costing implied by the various
+inter-dependencies and by the natural requirement that each step needs to be
+verifiable / diagnosable is as follows:
+
+* Import (rewrite) the relevant Linux headers elf.h and elftypes.h to managed
+  code, probably somewhere under *src/coreclr/src/tools/Common/Internal/Runtime*.
+  We need to verify licensing conditions for doing this. The actual work is
+  junior level, estimated at 2 man-weeks including some rudimentary regression
+  testing - e.g. writing a simple test that will dump a subset of the headers
+  for a hard-coded ELF file used as a resource by the test.
+  
+  An interesting observation is that most ELF header structures have
+  paraller 32-bit and 64-bit variants. It might be useful to logically unify them
```

Comment:
```suggestion
  parallel 32-bit and 64-bit variants. It might be useful to logically unify them
```

Reply:
Thanks, I actually noticed another typo, will fix both immediately

Reply:
OK, I cannot find "the other typo" right now. I fixed this one in the 2nd commit.